### PR TITLE
xbutil2 updates to report machine model, number of CPU cores as part …

### PR DESCRIPTION
**Pull request by sonals**
…of examine host

### CODE REVIEW

**Review by AShivangi — APPROVED** [review]

**Comment by sonals** [comment]
retest this please

File: src/runtime_src/core/edge/user/system_linux.cpp [hunk 1]
```diff
@@ -118,17 +144,19 @@ get_os_info(boost::property_tree::ptree &pt)
     }
     ifs.close();
   }
-
+  pt.put("model", machine_info());
+  pt.put("cores", std::thread::hardware_concurrency());
+  pt.put("memory", sysconf(_SC_PHYS_PAGES) * sysconf(_SC_PAGE_SIZE) / 0x100000);
```
Comment (rozumx) [inline src/runtime_src/core/edge/user/system_linux.cpp:149]: After reviewing the code base, I have come full circle to this entry.  Could you please do the following:

1. Change the "memory" entry name to "memory_bytes".
2. Remove the 0x100000 magic values.   It is best to keep the internal values in bytes, like we do with the other values.
3.  Add this value as a hex string value.  For example
`pt.put("memory_bytes", (boost::format("0x%x") %  sysconf(_SC_PHYS_PAGES) * sysconf(_SC_PAGE_SIZE).str().c_str())`

File: src/runtime_src/core/pcie/linux/system_linux.cpp [hunk 2]
```diff
@@ -133,6 +157,10 @@ get_os_info(boost::property_tree::ptree &pt)
       }
       ifs.close();
   }
+
+  pt.put("model", machine_info());
+  pt.put("cores", std::thread::hardware_concurrency());
+  pt.put("memory", sysconf(_SC_PHYS_PAGES) * sysconf(_SC_PAGE_SIZE) / 0x100000);
```
Comment (rozumx) [inline src/runtime_src/core/pcie/linux/system_linux.cpp:163]: Same comment as comment regarding changing the JSON property name and how the value is formatted.

File: src/runtime_src/core/pcie/windows/system_windows.cpp [hunk 3]
```diff
@@ -130,6 +131,17 @@ get_os_info(boost::property_tree::ptree &pt)
...
   pt.put("distribution", value);
+
+  BufferSize = sizeof value;
+  RegGetValueA(HKEY_LOCAL_MACHINE, "SYSTEM\\CurrentControlSet\\Control\\SystemInformation", "SystemProductName", RRF_RT_ANY, NULL, (PVOID)&value, &BufferSize);
+  pt.put("model", value);
+
+  MEMORYSTATUSEX mem;
+  mem.dwLength = sizeof(mem);
+  GlobalMemoryStatusEx(&mem);
+  pt.put("memory", mem.ullTotalPhys / 0x100000);
```
Comment (rozumx) [inline src/runtime_src/core/pcie/windows/system_windows.cpp:142]: Same comment as comment regarding changing the JSON property name and how the value is formatted.

File: src/runtime_src/core/tools/common/ReportHost.cpp [hunk 4]
```diff
@@ -70,16 +70,18 @@ ReportHost::writeReport(const xrt_core::device * _pDevice,
     _output << boost::format("  %-20s : %s\n") % "Release" % _pt.get<std::string>("host.os.release");
     _output << boost::format("  %-20s : %s\n") % "Version" % _pt.get<std::string>("host.os.version");
     _output << boost::format("  %-20s : %s\n") % "Machine" % _pt.get<std::string>("host.os.machine");
+    _output << boost::format("  %-20s : %s\n") % "CPU Cores" % _pt.get<std::string>("host.os.cores");
+    _output << boost::format("  %-20s : %s MB\n") % "Memory" % _pt.get<std::string>("host.os.memory");
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportHost.cpp:74]: O.K. - I now see that we are dealing with MBytes.

Comment (rozumx) [reply src/runtime_src/core/tools/common/ReportHost.cpp:74]: Being that the original data is coming in as a hex string, and it is desired to report it as MBs, the following code could be of assistance:
```
static const uint64_t BYTES_TO_MEGABYTES = 0x1000000;
_output << boost::format("  %-20s : 0x%x MB\n") % "Memory" % strtoll(_pt.get<std::string>("host.os.memory").c_str(),nullptr,16)/BYTES_TO_MEGABYTES);
```

File: src/runtime_src/core/pcie/linux/system_linux.cpp [hunk 5]
```diff
@@ -85,6 +86,29 @@ glibc_info()
   return _pt;
 }
 
+static std::string machine_info()
+{
+#if defined(__aarch64__) || defined(__arm__) || defined(__mips__)
```
Comment (rozumx) [inline src/runtime_src/core/pcie/linux/system_linux.cpp:91]: I have never been a fan of compiler #if def of variable declarations (they always are difficult to support).  By chance, could you refactor it to be something that would not require doing this.  For example:


```
#if defined(__aarch64__) || defined(__arm__) || defined(__mips__)
  #define MACHINE_NODE_PATH = "/proc/device-tree/model";
#elif defined(__PPC64__)
  #define MACHINE_NODE_PATH = "/proc/device-tree/model-name";
  // /proc/device-tree/system-id may be 000000
  // /proc/device-tree/model may be 00000
#elif defined (__x86_64__)
  #define MACHINE_NODE_PATH = "/sys/devices/virtual/dmi/id/product_name";
#else
  #error "Unsupported platform"
#endif

static std::string machine_info()
{
  static const char node[] = MACHINE_NODE_PATH;
  std::string model("unknown");
  std::ifstream stream(node);
  if (stream.good()) {
    std::getline(stream, model);
    stream.close();
  }
  return model;
}
```

Doing so keeps the keep readable easy to support.

**Review by rozumx — CHANGES_REQUESTED** [review]
In addition to the JSON name changes below, could you also add to the comments a sample output of the JSON report.  

Thanks

**Review by Himanshu-xilinx — APPROVED** [review]
looks good

File: src/runtime_src/core/edge/user/system_linux.cpp (hunk 1 above)
Comment (sonals) [reply src/runtime_src/core/edge/user/system_linux.cpp]: Changes made to store memory as as hex string.

File: src/runtime_src/core/pcie/windows/system_windows.cpp (hunk 3 above)
Comment (sonals) [reply src/runtime_src/core/pcie/windows/system_windows.cpp]: Done

File: src/runtime_src/core/tools/common/ReportHost.cpp (hunk 4 above)
Comment (sonals) [reply src/runtime_src/core/tools/common/ReportHost.cpp]: Thanks for the suggestion -- done!

File: src/runtime_src/core/pcie/linux/system_linux.cpp (hunk 5 above)
Comment (sonals) [reply src/runtime_src/core/pcie/linux/system_linux.cpp]: Done!

File: src/runtime_src/core/pcie/windows/system_windows.cpp [hunk 6]
```diff
@@ -174,7 +186,7 @@ get_mgmtpf_device(device::id_type id) const
 
 void
 system_windows::
-program_plp(std::shared_ptr<device> dev, std::vector<char> buffer) const 
+program_plp(std::shared_ptr<device> dev, std::vector<char> buffer) const
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/system_windows.cpp:191]: Not part of this PR, but still has to be fixed.  Do not pass by value, especially `buffer ` looks very inefficient, but also dev should be pass by const reference.

File: src/runtime_src/core/pcie/windows/system_windows.cpp [hunk 7]
```diff
@@ -29,6 +29,7 @@
 #include <chrono>
 #include <thread>
 #include <ctime>
+#include <thread>
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/system_windows.cpp:32]: Why this extra include?  already included line 30.

File: src/runtime_src/core/pcie/linux/system_linux.cpp [hunk 8]
```diff
@@ -85,6 +99,18 @@ glibc_info()
   return _pt;
 }
 
+static std::string machine_info()
+{
+  static const char node[] = MACHINE_NODE_PATH;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/system_linux.cpp:105]: What's the point of node here?  Why not use MACHINE_NODE_PATH directly line 106?

File: src/runtime_src/core/edge/user/system_linux.cpp [hunk 9]
```diff
@@ -165,7 +195,7 @@ get_mgmtpf_device(device::id_type id) const
 
 void
 system_linux::
-program_plp(std::shared_ptr<device> dev, std::vector<char> buffer) const 
+program_plp(std::shared_ptr<device> dev, std::vector<char> buffer) const
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/system_linux.cpp:199]: Not part of this PR, but still has to be fixed. Do not pass by value, especially buffer looks very inefficient, but also dev should be pass by const reference.

File: src/runtime_src/core/edge/user/system_linux.cpp [hunk 10]
```diff
@@ -85,7 +100,20 @@ glibc_info()
   return _pt;
 }
 
-void 
+static std::string
+machine_info()
+{
+  static const char node[] = MACHINE_NODE_PATH;
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/system_linux.cpp:107]: What's the point of node here? Why not use MACHINE_NODE_PATH directly line 108?

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/pcie/windows/system_windows.cpp (hunk 6 above)
Comment (sonals) [reply src/runtime_src/core/pcie/windows/system_windows.cpp]: I wonder how this crept in. I went ahead and updated the performance critical std::vector reference change `system::program_plp(std::shared_ptr<device> dev, const std::vector<char> &buffer)` 
Leaving out the shared_ptr device change for future.

File: src/runtime_src/core/pcie/windows/system_windows.cpp (hunk 7 above)
Comment (sonals) [reply src/runtime_src/core/pcie/windows/system_windows.cpp]: Removed

File: src/runtime_src/core/pcie/linux/system_linux.cpp (hunk 8 above)
Comment (sonals) [reply src/runtime_src/core/pcie/linux/system_linux.cpp]: Removed my brain dead artifact.

File: src/runtime_src/core/edge/user/system_linux.cpp (hunk 9 above)
Comment (sonals) [reply src/runtime_src/core/edge/user/system_linux.cpp]: Done, see above.

File: src/runtime_src/core/edge/user/system_linux.cpp (hunk 10 above)
Comment (sonals) [reply src/runtime_src/core/edge/user/system_linux.cpp]: Done, see above.

File: src/runtime_src/core/pcie/linux/system_linux.cpp (hunk 8 above)
Comment (rozumx) [reply src/runtime_src/core/pcie/linux/system_linux.cpp]: Yes, that will work as well.  

Personally for me, I like to use a static const variable to contain this type of information especially if the value is determine via the compiler environment.  Having this in a static const variable also enables one to easily inspect this value via the debugger, if needed in the future.  

Again, not a big deal.  It is just a coding pattern I like to do especially if it is being used as an argument into a function/method that is difficult to step into.

**Review by rozumx — APPROVED** [review]
Thank you for making the recommended changes.

File: src/runtime_src/core/edge/user/system_linux.cpp (hunk 9 above)
Comment (AShivangi) [reply src/runtime_src/core/edge/user/system_linux.cpp]: Sorry, that's on me. I will take care of this asap

File: src/runtime_src/core/pcie/linux/system_linux.cpp (hunk 8 above)
Comment (stsoe) [reply src/runtime_src/core/pcie/linux/system_linux.cpp]: Sounds like you suggest working around the misuse of #define :-)  
In reality all the #defines should be `constexpr char blah[] = "blahblah"`   
Should we outlaw #defines in C++ code?
